apps/hash-graph/lib/graph/src/store/postgres/knowledge/entity.rs

```diff
@@ -87,34 +85,48 @@ impl<C: AsClient> PostgresStore<C> {
-                    let entity_type_id =
-                        EntityTypeVertexId::from(entity.metadata.entity_type_id().clone());
-                    subgraph.insert_edge(
-                        &entity_vertex_id,
-                        SharedEdgeKind::IsOfType,
-                        entity_type_id.clone(),
-                    );
+                    for entity_type in <Self as Read<EntityTypeWithMetadata>>::read(
+                        self,
+                        &Filter::for_shared_edge_by_entity_id(
+                            entity_vertex_id.base_id,
+                            SharedEdgeKind::IsOfType,
+                        ),
+                        &temporal_axes,
+                    )
+                    .await?
+                    {
+                        let entity_type_vertex_id = entity_type.vertex_id(time_axis);
+
+                        subgraph.insert_edge(
+                            &entity_vertex_id,
+                            SharedEdgeKind::IsOfType,
+                            entity_type_vertex_id.clone(),
+                        );
+
+                        subgraph.insert_vertex(&entity_type_vertex_id, entity_type);
 
-                    self.traverse_entity_type(
-                        vec![entity_type_id],
-                        temporal_axes.clone(),
-                        GraphResolveDepths {
-                            is_of_type: OutgoingEdgeResolveDepth {
-                                outgoing: graph_resolve_depths.is_of_type.outgoing - 1,
-                                ..graph_resolve_depths.is_of_type
+                        entity_type_queue.push((
+                            entity_type_vertex_id,
+                            GraphResolveDepths {
+                                is_of_type: OutgoingEdgeResolveDepth {
+                                    outgoing: graph_resolve_depths.is_of_type.outgoing - 1,
+                                    ..graph_resolve_depths.is_of_type
+                                },
+                                ..graph_resolve_depths
                             },
-                            ..graph_resolve_depths
-                        },
-                        traversal_context,
-                        subgraph,
-                    )
-                    .await?;
+                            temporal_axes.clone(),
+                        ));
+                    }
                 }
 
                 if graph_resolve_depths.has_left_entity.incoming > 0 {
                     for outgoing_link_entity in <Self as Read<Entity>>::read(
                         self,
-                        &Filter::for_outgoing_link_by_source_entity_id(entity_vertex_id.base_id),
+                        &Filter::for_knowledge_graph_edge_by_entity_id(
+                            entity_vertex_id.base_id,
+                            KnowledgeGraphEdgeKind::HasLeftEntity,
+                            true,
```

Comment:
Is it worth wrapping these in a struct and creating aliases kind of like we've done in TypeScript (although obviously easier there)?

Just thinking it might be too easy to accidentally put the wrong boolean in or the wrong edge-kind, and it'd be nice to just say `OutgoingLinkEdge` or something.

Could be done in follow-up if you think it's worth it

Reply:
I replaced the boolean with an enumeration in #2226

Reply:
Yeah I noticed that :) 

I think it'd still be nice to know that an `OutgoingLink` is equivalent to `HasLeftEntity, Incoming`

Reply:
Similar to the aliases in this file:
https://github.com/blockprotocol/blockprotocol/blob/fc2153c3d86d53f10aa76108360a0ea35e98f16f/libs/%40blockprotocol/graph/src/shared/types/subgraph/edges/variants/knowledge.ts#L8C20-L15

Reply:
I prefer to not add those aliases in this location/at this point. We don't consume the subgraph, currently, and it's only partially the truth. The subgraph does not imply meaning to link directions (thus, it even supports right-to-left links bi-directional links)
Wrapping those in a struct would be a step back as I need exactly this representation to unify the interface and make a subgraph, which can resolve only on edge kinds. If these would be in a struct again, this would need to be unwrapped or a trait would need to be defined.

Reply:
It's fixed now: [Fix traversal direction](https://github.com/hashintel/hash/pull/2223/commits/de548a6431d89fd2147ba9a6e9456f7326db7251)

Reply:
I think it would help readability a _**lot**_, otherwise we are going to easily run into problems exactly like the ones described below.

> make a subgraph, which can resolve only on edge kinds

In that case wouldn't we want something even more like this, to make sure that seeing some random edge kind and a direction would be understandable.

I guess the main issue is that we can't just wrap the two parameters (KnowledgeGraphEdgeKind, bool) into one generic struct and write aliases, because we can't constrain it to a specific enum variant? If there isn't an easy way around that then yea

Reply:
Except something like this _would_ be more readable, I just appreciate we may not easily be able to achieve it.

```rust
 for (edge_kind, edge_direction) in &[ 
     OutgoingLinkEdge, 
     IncomingLinkEdge 
     LeftEntityEdge, 
     RightEntityEdge, 
 ] { 
     entity_queue.extend( 
         self.traverse_knowledge_graph_edge( 
             entity_vertex_id, 
             entity_interval, 
             *edge_kind, 
             *edge_direction, 
             graph_resolve_depths, 
             &temporal_axes, 
             traversal_context, 
             subgraph, 
         ) 
         .await?, 
     ); 
 } 
```

Reply:
But what is an `OutgoingLinkEdge`? Either, we introduce such an edge or it would be confusing as soon as we have the non-default case, where the left entity is the target of the right entity.

Reply:
As I expressed offline, we have an `OutgoingLinkEdge`, it is a thing in the official implementation of the spec. This is an implementation of a datastore that respects that spec.

Should the spec be updated to allow more things, we can always adjust the code.

> But what is an OutgoingLinkEdge

I don't really understand why this is so controversial when _you've already written it in the code_
https://github.com/hashintel/hash/blob/a30bc9e4ea3b8b83dc0b5cdefa9d7c16256391da/apps/hash-graph/lib/graph/src/store/postgres/knowledge/entity.rs#L123

Reply:
If you feel really strongly that there shouldn't be such an alias as `OutgoingLink` then a separate discussion can be had about that within the context of the Block Protocol and the spec, but at the moment it _is_ a thing.

Reply:
Because the above abstracts over this. Also, this does not align with the idea of completely abstracting away the logic. We **_don't_** have an `OutgoingLinkEdge` in the resolve depths and the traversal logic should only take the resolve depths into account when traversing it.

> Could be done in follow-up if you think it's worth it

I don't think it's worth it.

Reply:
> If you feel really strongly that there shouldn't be such an alias as OutgoingLink then a separate discussion can be had about that within the context of the Block Protocol and the spec, but at the moment it is a thing.

Let's discuss this offline but please let us not block this PR on this as the code you commented on is already merged.